### DISCUSSION TITLE
Download queue feature

### DISCUSSION THREAD

**Discussion by vanossj** [lassoan/SlicerDICOMwebBrowser#22] [Ideas]
I often want to download a couple series from multiple studies.

But if I start downloading series from one study, the UI is locked and I can't select a new study or start downloading more series until the first set of series has finished downloading (and added to the database)

One potential solution would be to add a download queue, you could add series to the queue, and then you would have to press a button to actually start downloading them.

Looking for feedback on this idea or alternative solutions before I start coding anything

**Comment by lassoan** [comment]
One possitibility would be to use https://github.com/pieper/SlicerParallelProcessing extension. It provides a framework for performing background tasks in separate processes. An example processing task:
- [Process class running in Slicer](https://github.com/pieper/SlicerParallelProcessing/blob/c9f2a11e1b51ea28d40b40c7a04489eeb2fe6341/Processes/Processes.py#L309-L338) that prepares the input (in your case the inputs are the connection information, the UIDs you want to retrieve, and the folder you want to save them) and uses the output (in your case it would be import the downloaded items into the DICOM database and notify the DICOM/DICOMweb browser)
- [Processing script](https://github.com/pieper/SlicerParallelProcessing/blob/master/Processes/Resources/ProcessScripts/modelFilter.slicer.py) that runs in the background process: in your case it would download the files using dicomweb-client
- [Example of how to set up and run processing](https://github.com/pieper/SlicerParallelProcessing/blob/c9f2a11e1b51ea28d40b40c7a04489eeb2fe6341/Processes/Processes.py#L357-L383)

One current limitation of this framework that there is no progress reporting from within a task ([you only get a "finished" signal](https://github.com/pieper/SlicerParallelProcessing/blob/c9f2a11e1b51ea28d40b40c7a04489eeb2fe6341/Processes/Processes.py#L226) when the processing task is completed). It could be possible to add a mechanism that the processing scipt would print progress information (e.g., json string) on its outout and the Process class in Slicer could look at the process output time-to-time and if it finds progress information in it then it would forward it to the GUI.

We are also working on a new-generation DICOM browser ("visual DICOM browser") in Slicer that will allow query/retrieve for remote servers and displaying results in a nice layout, with thumbnails. This is implemented in C++ and uses classic DIMSE networking but it would be nice to add DICOMweb support for it. Although it is a C++ class, actual query/retrieve calls could be implemented by calling Python code, similarly to how it is done in Parallel Processing module. So, downloading data in separate Python processes could be a good preparatory work for this.

**Reply by lassoan** [reply]
It could also be possible to emulate background downloading on a single thread by downloading instances from a timer callback function (instead in a loop). Or even Python threads might be usable, as the main time sink is waiting for network I/O completion.
